packages/aws-cdk/lib/api/garbage-collection/garbage-collector.ts

```diff
@@ -0,0 +1,389 @@
+}
+
+/**
+ * A class to facilitate Garbage Collection of S3 and ECR assets
+ */
+export class GarbageCollector {
+  private garbageCollectS3Assets: boolean;
+  private garbageCollectEcrAssets: boolean;
+  private permissionToDelete: boolean;
+  private permissionToTag: boolean;
+  private bootstrapStackName: string;
+  private maxWaitTime: number;
+
+  public constructor(readonly props: GarbageCollectorProps) {
+    this.garbageCollectS3Assets = ['s3', 'all'].includes(props.type);
+    this.garbageCollectEcrAssets = ['ecr', 'all'].includes(props.type);
+
+    debug(`${this.garbageCollectS3Assets} ${this.garbageCollectEcrAssets}`);
+
+    this.permissionToDelete = ['delete-tagged', 'full'].includes(props.action);
+    this.permissionToTag = ['tag', 'full'].includes(props.action);
+    this.maxWaitTime = props.maxWaitTime ?? 60000;
+
+    this.bootstrapStackName = props.bootstrapStackName ?? DEFAULT_TOOLKIT_STACK_NAME;
+
+    // TODO: ECR garbage collection
+    if (this.garbageCollectEcrAssets) {
+      throw new Error('ECR garbage collection is not yet supported');
+    }
+  }
+
+  /**
+   * Perform garbage collection on the resolved environment.
+   */
+  public async garbageCollect() {
+    // SDKs
+    const sdk = (await this.props.sdkProvider.forEnvironment(this.props.resolvedEnvironment, Mode.ForWriting)).sdk;
+    const cfn = sdk.cloudFormation();
+    const s3 = sdk.s3();
+
+    const qualifier = await this.bootstrapQualifier(sdk, this.bootstrapStackName);
+    const activeAssets = new ActiveAssetCache();
+
+    // Grab stack templates first
+    const startTime = Date.now();
+    await refreshStacks(cfn, activeAssets, this.maxWaitTime, qualifier);
+    // Start the background refresh
+    const backgroundStackRefresh = new BackgroundStackRefresh({
+      cfn,
+      activeAssets,
+      qualifier,
+      maxWaitTime: this.maxWaitTime,
+    });
+    const timeout = setTimeout(backgroundStackRefresh.start, Math.max(startTime + 300_000 - Date.now(), 0));
+
+    const bucket = await this.bootstrapBucketName(sdk, this.bootstrapStackName);
+    const numObjects = await this.numObjectsInBucket(s3, bucket);
+    const printer = new ProgressPrinter(numObjects, 1000);
+
+    debug(`Found bootstrap bucket ${bucket}`);
+
+    try {
+      const batches = 1;
+      const batchSize = 1000;
+      const currentTime = Date.now();
+      const graceDays = this.props.rollbackBufferDays;
+
+      debug(`Parsing through ${numObjects} objects in batches`);
+
+      // Process objects in batches of 1000
+      // This is the batch limit of s3.DeleteObject and we intend to optimize for the "worst case" scenario
+      // where gc is run for the first time on a long-standing bucket where ~100% of objects are isolated.
+      for await (const batch of this.readBucketInBatches(s3, bucket, batchSize, currentTime)) {
+        print(chalk.green(`Processing batch ${batches} of ${Math.floor(numObjects / batchSize) + 1}`));
+        printer.start();
```

Comment:
I think we need the following in here, to protect us from problems in the refresher:

```ts
await backgroundStackRefresh.noOlderThan(10 * MINUTE_MS);
```



To be implemented roughly like:

```ts
class BackgroundRefresher {
  private lastRefreshTime: number;
  private queuedPromises: Array<() => void>;

  public noOlderThan(ms: number) {
    const horizon = Date.now() - ms;
    if (this.lastRefreshTime >= horizon) { return Promise.resolve(); }
    
    return Promise.race([
      new Promise(resolve => this.queuedPromises.push(resolve)),
      new Promise((_, reject) => setTimeout(() => reject(new Error('Took too long, the background thread probably exploded or summin')), ms)),
    ]);
  }

   private justRefreshedStacks() {
     this.lastRefreshTime = Date.now();
     for (const p of this.queuedPromises.splice(0, this.queuedPromises.length)) {
       p();
     }
}
```